Setup.hs: Updates for ghc-8.2 and Cabal 2.0

erikd-ambiata opened the pull request:
Tested with GHC 7.10, 8.0 and 8.2.
    
Because this script is run with `runhaskell` it is not possible
to use the normal cabal `MIN_VERSION_<pkg>` macros and have to
rely on GHC version information instead.

! @charleso @jystic @nhibberd 
/jury approved @charleso

charleso commented:
👍 

Would want to check with Jacob/Nick too

erikd-ambiata commented:
Problem is actually more subtle than this PR suggests. Needs more work.

erikd-ambiata commented:
Fixed now. I've also put this version to work in parsley. See  https://github.com/ambiata/parsley/pull/37

Won't merge this until all three of you give it 👍 .

jystic commented on src/Setup.hs:
```diff
@@ -4,17 +4,29 @@ import           Data.Char (isDigit, toLower)
+--
+--    custom-setup
+--        setup-depends:       base
+--                           , Cabal >= 2.0
```
how does mafia cope with the `custom-setup` section? I know it's caused some problems in the past is all

erikd-ambiata replied:
It copes as long as cabal-install is 1.24 or later. Previous problems were probably cabal-install 1.22 which I have confirmed not to work.

jystic commented:
> Because this script is run with `runhaskell` it is not possible to use the normal cabal `MIN_VERSION_<pkg>` macros and have to rely on GHC version information instead.

Is this the case with GHC 8? I thought that GHC itself generates these macros from v8 onwards

erikd-ambiata commented:
> Is this the case with GHC 8? I thought that GHC itself generates these macros from v8 onwards

Maybe. Main things is that we want to build with any combination of ghc 7.10/8.0 and cabal-install 1.24/2.0.

erikd-ambiata commented:
This still doesn't work in all situations. Trying to figure out why.

erikd-ambiata committed:
Fix Setup.hs

This version of Setup.hs will work with any of the three versions
of GHC (7.10, 8.0 and 8.2) and either of two versions of cabal-install
(1.24 and 2.0).

It will however break with ghc-8.4 (due Febrary 2018) due to yet
another change in the API of the Cabal library.

erikd-ambiata commented:
The version in commit 0292d0e10f1fe830f5e6e9692587a873af19355c is the same as the one that is currently in Walrus.

charleso commented:
@erikd-ambiata I'm having problems using this locally:

```
> mafia build
Failed to run 'cabal sdist' for source package: ../walrus-conduit
Process failed with exit code: 1
[1 of 1] Compiling Main             ( dist/setup/setup.hs, dist/setup/Main.o )
dist/setup/setup.hs:1:2: error: parse error on input ‘#!/’
```

```
> ghc --version
The Glorious Glasgow Haskell Compilation System, version 8.0.1
> cabal --version
cabal-install version 1.24.0.1
compiled using version 1.24.1.0 of the Cabal library
```

If I take out the first line of the `Setup.hs` file it work (eg. take out the shebang).

erikd-ambiata commented:
That's weird, because that file works in Walrus. 

I do notice that in Walrus the file has execute permissions:
```
> ls framework/Setup.hs 
-rwxr-x--- 1 erikd erikd 5188 Jan  3 14:33 framework/Setup.hs
```
What about yours?

charleso commented:
https://github.com/haskell/haskell-ide-engine/issues/333

charleso commented:
(It's an OSX thing apparently)

erikd-ambiata commented:
Actual issue is the GHC on OSX uses Clang as the C pre-processor whereas on Linux GCC will be used.
